eppy/bunch_subclass.py

```diff
@@ -119,73 +143,73 @@ def addfunctions(abunch):
         "WINDOWMATERIAL:SCREEN",
         "WINDOWMATERIAL:SHADE",
         "WINDOWMATERIAL:SIMPLEGLAZINGSYSTEM",
-              ]
-    if key in names:
-        func_dict = {
-            'rvalue': fh.rvalue,
-            'ufactor': fh.ufactor,
-            'rvalue_ip': fh.rvalue_ip,  # quick fix for Santosh. Needs to thought thru
-            'ufactor_ip': fh.ufactor_ip,  # quick fix for Santosh. Needs to thought thru
-            'heatcapacity': fh.heatcapacity,
-        }
-        abunch.__functions.update(func_dict)
+    ]
+    func_dict = {
+        'rvalue': (fh.rvalue, names),
+        'ufactor': (fh.ufactor, names),
+        'rvalue_ip': (fh.rvalue_ip, names),
+        # quick fix for Santosh. Needs to thought thru
+        'ufactor_ip': (fh.ufactor_ip, names),
+        # quick fix for Santosh. Needs to thought thru
+        'heatcapacity': (fh.heatcapacity, names),
+    }
+    abunch._EpBunch__functions.update(func_dict)
 
     names = [
         'FAN:CONSTANTVOLUME',
         'FAN:VARIABLEVOLUME',
         'FAN:ONOFF',
         'FAN:ZONEEXHAUST',
         'FANPERFORMANCE:NIGHTVENTILATION',
-              ]
-    if key in names:
-        func_dict = {
-            'f_fanpower_bhp': fh.fanpower_bhp,
-            'f_fanpower_watts': fh.fanpower_watts,
-            'f_fan_maxcfm': fh.fan_maxcfm,
-        }
-        abunch.__functions.update(func_dict)
+    ]
+    func_dict = {
+        'f_fanpower_bhp': (fh.fanpower_bhp, names),
+        'f_fanpower_watts': (fh.fanpower_watts, names),
+        'f_fan_maxcfm': (fh.fan_maxcfm, names),
+    }
+    abunch._EpBunch__functions.update(func_dict)
     # =====
     # code for references
-    #-----------------
+    # -----------------
     # add function zonesurfaces
     if key == 'ZONE':
-        func_dict = {'zonesurfaces':fh.zonesurfaces}
-        abunch.__functions.update(func_dict)
+        func_dict = {'zonesurfaces': (fh.zonesurfaces, ['ZONE'])}
+        abunch._EpBunch__functions.update(func_dict)
 
-    #-----------------
+    # -----------------
     # add function subsurfaces
     # going to cheat here a bit
     # check if epbunch has field "Zone_Name"
     # and is in group u'Thermal Zones and Surfaces'
     # then it is likely to be a surface attached to a zone
     fields = abunch.fieldnames
-    try:
-        group = abunch.getfieldidd('key')['group']
-    except KeyError as e:  # some pytests don't have group
-        group = None
-    if group == u'Thermal Zones and Surfaces':
-        if "Zone_Name" in fields:
-            func_dict = {'subsurfaces':fh.subsurfaces}
-            abunch.__functions.update(func_dict)
+    names = ['BuildingSurface:Detailed', 'Wall:Detailed',
```

Comment:
I don't like this hardcoded list. What if more surface types are added to EnergyPlus?

Reply:
I agree! Is there a IDD group for subsurfaces?